ic-log/src/writer.rs

```diff
@@ -46,43 +48,85 @@ const INIT_LOG_CAPACITY: usize = 128;
 
 type LogRecordsBuffer = AllocRingBuffer<String>;
 thread_local! {
-    static LOG_RECORDS: RefCell<LogRecordsBuffer> =
-        RefCell::new(LogRecordsBuffer::new(INIT_LOG_CAPACITY));
+    static LOG_RECORDS: RefCell<(usize, LogRecordsBuffer)> =
+        RefCell::new((0, LogRecordsBuffer::new(INIT_LOG_CAPACITY)));
 }
 
 /// Writer that stores strings in a thread_local memory circular buffer.
 /// Note: it can be optimized to reduce the number of memory allocations.
 pub struct InMemoryWriter {}
 
+#[derive(Debug, Default, PartialEq, Eq, CandidType, Serialize, Deserialize)]
+pub struct Logs {
+    /// the list of logs
+    pub logs: Vec<Log>,
+    /// the count of available logs
+    pub all_logs_count: usize,
+}
+
+#[derive(Debug, Default, PartialEq, Eq, CandidType, Serialize, Deserialize)]
+pub struct Log {
+    /// the log text
+    pub log: String,
+    /// the offset of the log
+    pub offset: usize,
+}
+
 impl InMemoryWriter {
     pub fn init_buffer(capacity: usize) {
         LOG_RECORDS.with(|records| {
-            *records.borrow_mut() = LogRecordsBuffer::new(capacity);
+            *records.borrow_mut() = (0, LogRecordsBuffer::new(capacity));
         });
     }
 
-    pub fn take_records(max_count: usize) -> Vec<String> {
+    pub fn take_records(max_count: usize, from_offset: usize) -> Logs {
         LOG_RECORDS.with(|records| {
-            let mut records = records.borrow_mut();
-            let mut result = Vec::with_capacity(max_count);
-            for _ in 0..max_count {
-                if let Some(s) = records.dequeue() {
-                    result.push(s);
+            let records = records.borrow_mut();
+            let all_logs_count = records.0;
+
+            if (from_offset >= all_logs_count) || all_logs_count == 0 {
+                Logs {
+                    all_logs_count,
+                    logs: vec![],
+                }
+            } else {
+                let mut result = Vec::with_capacity(max_count);
+
+                let from_offset = if (from_offset + records.1.len()) < all_logs_count {
```

Comment:
maybe I would move this block to a separate function and test it standalone, because it's a little hard to read